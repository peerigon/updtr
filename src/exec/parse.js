const STRING_PROPERTIES = ["name", "current", "wanted", "latest"];

/**
 * Applies JSON.parse on the str.
 * Additionally, returns null if the str was empty or just contained whitespace.
 *
 * @param {string} str
 * @returns {*}
 */
function parse(str) {
    const trimmed = str.trim();

    if (trimmed.length === 0) {
        return null;
    }

    let resultJSON = {};
    try {
        resultJSON = JSON.parse(trimmed);
    } catch (e) {
        const eol = require("os").EOL;
        const lines = trimmed.split(eol);
        lines.map(stringJSON => {
            const possible = JSON.parse(stringJSON);
            if (possible.type === "table") {
                resultJSON = possible;
            }
        })
    }

    return resultJSON;
}

function returnIfValid(result) {
    STRING_PROPERTIES.forEach(prop => {
        if (typeof result[prop] !== "string") {
            throw new Error("Unexpected output format of package manager");
        }
    });

    return result;
}

function tryParse(normalizer) {
    return (stdout, cmd) => {
        try {
            return normalizer(parse(stdout));
        } catch (err) {
            err.message = `Error when trying to parse stdout from command '${ cmd }': ${ err.message }`;
            throw err;
        }
    };
}

function arrToObj(arr, keys) {
    return keys.reduce((obj, key, i) => {
        obj[key] = arr[i];

        return obj;
    }, {});
}

// By sorting the parsed data, we get deterministic results across different npm and yarn versions.
// As a nice side-effect, a package like eslint will always be updated before eslint-config-peerigon
// which might have a peer dependency on eslint
// See https://github.com/peerigon/updtr/issues/48
function sortByName(o1, o2) {
    return o1.name > o2.name;
}

const list = tryParse(
    parsed =>
        (parsed.dependencies === undefined ?
            [] :
            Object.keys(parsed.dependencies)
                .map(name => ({
                    name,
                    version: parsed.dependencies[name].version,
                }))
                .sort(sortByName))
);

export default {
    npm: {
        outdated: tryParse(parsed => {
            if (parsed === null) {
                return [];
            }

            const names = Object.keys(parsed);

            return names
                .map(name => parsed[name])
                .map((dep, index) =>
                    returnIfValid({
                        name: names[index],
                        current: dep.current,
                        wanted: dep.wanted,
                        latest: dep.latest,
                    })
                )
                .sort(sortByName);
        }),
        list,
    },
    yarn: {
        outdated: tryParse(
            parsed =>
                (parsed === null ?
                    [] :
                    parsed.data.body
                        .map(row => arrToObj(row, parsed.data.head))
                        .map(dep =>
                            returnIfValid({
                                name: dep.Package,
                                current: dep.Current,
                                wanted: dep.Wanted,
                                latest: dep.Latest,
                            })
                        )
                        .sort(sortByName))
        ),
        // We currently only use npm for the list command (see cmds.js).
        // Put the real implementation here if we decide to use yarn for list.
        list,
    },
};
