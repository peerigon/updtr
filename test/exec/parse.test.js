import readFixtures from "../helpers/readFixtures";
import parse from "../../src/exec/parse";

let jsonUnexpectedEndOfInput;
let stdoutLogs;

function testUnexpectedInput(parse) {
    describe("unexpected input", () => {
        describe("malformed JSON", () => {
            it("should throw with a helpful error message", () => {
                expect(() => parse("{", "my-cmd")).toThrow(
                    "Error when trying to parse stdout from command 'my-cmd': " +
                        jsonUnexpectedEndOfInput.message
                );
            });
        });
        describe("unexpected data format", () => {
            it("should throw with a helpful error message", () => {
                expect(() =>
                    parse(
                        JSON.stringify({ some: { randomData: true } }),
                        "my-cmd"
                    )
                )
                    // We don't really want to test for the actual error message here,
                    // but just whether there is an error message at all.
                    .toThrow(
                    /Error when trying to parse stdout from command 'my-cmd': \w+/
                    );
            });
        });
    });
}

beforeAll(async () => {
    stdoutLogs = await readFixtures([
        "empty/outdated.npm.log",
        "empty/outdated.yarn.log",
        "empty/list.npm.log",
        "empty/list.yarn.log",
        "no-outdated/outdated.npm.log",
        "no-outdated/outdated.yarn.log",
        "no-outdated/list.npm.log",
        "no-outdated/list.yarn.log",
        "no-outdated-dev/outdated.npm.log",
        "no-outdated-dev/outdated.yarn.log",
        "no-outdated-dev/list.npm.log",
        "no-outdated-dev/list.yarn.log",
        "outdated/outdated.npm.log",
        "outdated/outdated.yarn.log",
        "outdated/list.npm.log",
        "outdated/list.yarn.log",
        "outdated-dev/outdated.npm.log",
        "outdated-dev/outdated.yarn.log",
        "outdated-dev/list.npm.log",
        "outdated-dev/list.yarn.log",
    ]);
    try {
        JSON.parse("");
    } catch (err) {
        jsonUnexpectedEndOfInput = err;
    }
});

describe("parse", () => {
    describe(".npm", () => {
        describe(".outdated()", () => {
            describe("empty fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("empty/outdated.npm.log")
                        .trim();

                    expect(fixture.length).toBe(0);
                    expect(() => JSON.parse(fixture)).toThrow(
                        jsonUnexpectedEndOfInput
                    );
                });
                it("should return an empty array", () => {
                    expect(
                        parse.npm.outdated(
                            stdoutLogs.get("empty/outdated.npm.log")
                        )
                    ).toEqual([]);
                });
            });
            describe("no-outdated fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("no-outdated/outdated.npm.log")
                        .trim();

                    expect(fixture.length).toBe(0);
                    expect(() => JSON.parse(fixture)).toThrow(
                        jsonUnexpectedEndOfInput
                    );
                });
                it("should return an empty array", () => {
                    expect(
                        parse.npm.outdated(
                            stdoutLogs.get("no-outdated/outdated.npm.log")
                        )
                    ).toEqual([]);
                });
            });
            describe("no-outdated-dev fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("no-outdated-dev/outdated.npm.log")
                        .trim();

                    expect(fixture.length).toBe(0);
                    expect(() => JSON.parse(fixture)).toThrow(
                        jsonUnexpectedEndOfInput
                    );
                });
                it("should return an empty array", () => {
                    expect(
                        parse.npm.outdated(
                            stdoutLogs.get("no-outdated-dev/outdated.npm.log")
                        )
                    ).toEqual([]);
                });
            });
            describe("outdated fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("outdated/outdated.npm.log")
                        .trim();

                    expect(fixture.length).toBeGreaterThan(10);
                    JSON.parse(fixture); // should be parseable
                });
                it("should return an array with normalized outdated data", () => {
                    expect(
                        parse.npm.outdated(
                            stdoutLogs.get("outdated/outdated.npm.log")
                        )
                    ).toMatchSnapshot();
                });
            });
            describe("outdated-dev fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("outdated-dev/outdated.npm.log")
                        .trim();

                    expect(fixture.length).toBeGreaterThan(10);
                    JSON.parse(fixture); // should be parseable
                });
                it("should return an array with normalized outdated data", () => {
                    expect(
                        parse.npm.outdated(
                            stdoutLogs.get("outdated-dev/outdated.npm.log")
                        )
                    ).toMatchSnapshot();
                });
            });
            testUnexpectedInput(parse.npm.outdated);
        });
        describe(".list()", () => {
            function testFixture(fixtureName) {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get(fixtureName + "/list.npm.log")
                        .trim();

                    expect(fixture.length).toBeGreaterThan(10);
                    JSON.parse(fixture); // should be parseable
                });
                it("should return an array with installed versions", () => {
                    expect(
                        parse.npm.list(
                            stdoutLogs.get(fixtureName + "/list.npm.log")
                        )
                    ).toMatchSnapshot();
                });
            }

            describe("empty fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs.get("empty/list.npm.log").trim();

                    expect(fixture.length).toBeGreaterThan(10);
                    JSON.parse(fixture); // should be parseable
                });
                it("should return an empty array", () => {
                    expect(
                        parse.npm.list(stdoutLogs.get("empty/list.npm.log"))
                    ).toEqual([]);
                });
            });

            describe("no-outdated fixture", () => {
                testFixture("no-outdated");
            });

            describe("no-outdated-dev fixture", () => {
                testFixture("no-outdated-dev");
            });

            describe("outdated fixture", () => {
                testFixture("outdated");
            });

            describe("outdated-dev fixture", () => {
                testFixture("outdated-dev");
            });
        });
    });
    describe(".yarn", () => {
        describe(".outdated()", () => {
            describe("empty fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("empty/outdated.yarn.log")
                        .trim();

                    expect(fixture.length).toBe(0);
                    expect(() => JSON.parse(fixture)).toThrow(
                        jsonUnexpectedEndOfInput
                    );
                });
                it("should return an empty array", () => {
                    expect(
                        parse.yarn.outdated(
                            stdoutLogs.get("empty/outdated.yarn.log")
                        )
                    ).toEqual([]);
                });
            });
            describe("no-outdated fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("no-outdated/outdated.yarn.log")
                        .trim();

                    expect(fixture.length).toBe(0);
                    expect(() => JSON.parse(fixture)).toThrow(
                        jsonUnexpectedEndOfInput
                    );
                });
                it("should return an empty array", () => {
                    expect(
                        parse.yarn.outdated(
                            stdoutLogs.get("no-outdated/outdated.yarn.log")
                        )
                    ).toEqual([]);
                });
            });
            describe("no-outdated-dev fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("no-outdated-dev/outdated.yarn.log")
                        .trim();

                    expect(fixture.length).toBe(0);
                    expect(() => JSON.parse(fixture)).toThrow(
                        jsonUnexpectedEndOfInput
                    );
                });
                it("should return an empty array", () => {
                    expect(
                        parse.yarn.outdated(
                            stdoutLogs.get("no-outdated-dev/outdated.yarn.log")
                        )
                    ).toEqual([]);
                });
            });
            describe("outdated fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("outdated/outdated.yarn.log")
                        .trim();

                    expect(fixture.length).toBeGreaterThan(10);
                    JSON.parse(fixture); // should be parseable
                });
                it("should return an array with normalized outdated data", () => {
                    expect(
                        parse.yarn.outdated(
                            stdoutLogs.get("outdated/outdated.yarn.log")
                        )
                    ).toMatchSnapshot();
                });
            });
            describe("outdated-dev fixture", () => {
                it("fixture sanity test", () => {
                    const fixture = stdoutLogs
                        .get("outdated-dev/outdated.yarn.log")
                        .trim();

                    expect(fixture.length).toBeGreaterThan(10);
                    JSON.parse(fixture); // should be parseable
                });
                it("should return an array with normalized outdated data", () => {
                    expect(
                        parse.yarn.outdated(
                            stdoutLogs.get("outdated-dev/outdated.yarn.log")
                        )
                    ).toMatchSnapshot();
                });
            });
            testUnexpectedInput(parse.yarn.outdated);
        });
        describe(".list()", () => {
            it("should be the same implementation as .npm.list", () => {
                expect(parse.yarn.list).toBe(parse.npm.list);
            });
        });
    });
});
