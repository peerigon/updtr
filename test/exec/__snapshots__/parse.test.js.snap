// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse .npm .list() no-outdated fixture should return an array with installed versions 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "version": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "version": "2.1.1",
  },
]
`;

exports[`parse .npm .list() no-outdated-dev fixture should return an array with installed versions 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "version": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "version": "2.1.1",
  },
]
`;

exports[`parse .npm .list() outdated fixture should return an array with installed versions 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "version": "1.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "version": "2.0.0",
  },
]
`;

exports[`parse .npm .list() outdated-dev fixture should return an array with installed versions 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "version": "1.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "version": "2.0.0",
  },
]
`;

exports[`parse .npm .outdated() outdated fixture should return an array with normalized outdated data 1`] = `
Array [
  Object {
    "current": "1.0.0",
    "latest": "2.0.0",
    "name": "updtr-test-module-1",
    "wanted": "1.1.1",
  },
  Object {
    "current": "2.0.0",
    "latest": "2.1.1",
    "name": "updtr-test-module-2",
    "wanted": "2.1.1",
  },
]
`;

exports[`parse .npm .outdated() outdated-dev fixture should return an array with normalized outdated data 1`] = `
Array [
  Object {
    "current": "1.0.0",
    "latest": "2.0.0",
    "name": "updtr-test-module-1",
    "wanted": "1.1.1",
  },
  Object {
    "current": "2.0.0",
    "latest": "2.1.1",
    "name": "updtr-test-module-2",
    "wanted": "2.1.1",
  },
]
`;

exports[`parse .yarn .outdated() outdated fixture should return an array with normalized outdated data 1`] = `
Array [
  Object {
    "current": "1.0.0",
    "latest": "2.0.0",
    "name": "updtr-test-module-1",
    "wanted": "1.0.0",
  },
  Object {
    "current": "2.0.0",
    "latest": "2.1.1",
    "name": "updtr-test-module-2",
    "wanted": "2.0.0",
  },
]
`;

exports[`parse .yarn .outdated() outdated-dev fixture should return an array with normalized outdated data 1`] = `
Array [
  Object {
    "current": "1.0.0",
    "latest": "2.0.0",
    "name": "updtr-test-module-1",
    "wanted": "1.0.0",
  },
  Object {
    "current": "2.0.0",
    "latest": "2.1.1",
    "name": "updtr-test-module-2",
    "wanted": "2.0.0",
  },
]
`;
